File: NeuralNets/Feedforward/typescript/tensorflow.js/src/utils/NeuralNet.ts
import * as tf from "@tensorflow/tfjs";

export class NeuralNet {
  NPL: number[];
  layers: number;
  learningRate: number;
  activations: tf.Tensor[];
  weights: tf.Tensor[];
  weightDerivatives: tf.Tensor[];
  biases: tf.Tensor[];
  biasDerivatives: tf.Tensor[];

  constructor(NPL: number[], learningRate: number = 0.01) {
    this.NPL = NPL;
    this.layers = NPL.length;
    this.learningRate = learningRate;
    this.activations = [];
    this.weights = [];
    this.weightDerivatives = [];
    this.biases = [];
    this.biasDerivatives = [];

    for (let currLayer = 0; currLayer < this.layers - 1; currLayer++) {
      this.activations.push(tf.zeros([this.NPL[currLayer]]));
      this.weights.push(
        tf.randomUniform(
          [this.NPL[currLayer], this.NPL[currLayer + 1]],
          -0.3,
          0.3
        )
      );
      this.weightDerivatives.push(
        tf.zeros([this.NPL[currLayer], this.NPL[currLayer + 1]])
      );
      this.biases.push(tf.zeros([this.NPL[currLayer + 1]]));
      this.biasDerivatives.push(tf.zeros([this.NPL[currLayer + 1]]));
    }

    this.activations.push(tf.zeros([this.NPL[this.layers - 1]]));
  }

  train(X: tf.Tensor[], Y: tf.Tensor[], epochs: number): void {
    const trainingData: [tf.Tensor, tf.Tensor][] = X.map((value, index) => [
      value,
      Y[index],
    ]);
    for (let epoch = 0; epoch < epochs; epoch++) {
      let meanLoss: number = 0;
      for (let [trainingInput, expectedOutput] of trainingData) {
        this.activations[0] = trainingInput;

        this.forwardprop();
        this.backprop(expectedOutput);
        this.updateNetwork();

        meanLoss += this.calculateLoss(expectedOutput);
      }

      meanLoss /= X.length;
      console.log(`Epoch[${epoch}]: Loss = ${meanLoss}`);
    }
  }

  forwardprop(): void {
    for (let currLayer = 0; currLayer < this.layers - 1; currLayer++) {
      this.activations[currLayer + 1] = tf.sigmoid(
        tf.add(
          tf.dot(this.activations[currLayer], this.weights[currLayer]),
          this.biases[currLayer]
        )
      );
    }
  }

  backprop(expectedOutput: tf.Tensor): void {
    let error: tf.Tensor = tf.sub(
      expectedOutput,
      this.activations[this.layers - 1]
    );

    for (let currLayer = this.layers - 1; currLayer > 0; currLayer--) {
      let delta: tf.Tensor = tf.mul(
        error,
        tf.mul(
          this.activations[currLayer],
          tf.sub(1, this.activations[currLayer])
        )
      );

      this.weightDerivatives[currLayer - 1] = tf.outerProduct(
        this.activations[currLayer - 1] as tf.Tensor1D,
        delta as tf.Tensor1D
      );
      this.biasDerivatives[currLayer - 1] = delta;

      error = tf.dot(delta, tf.transpose(this.weights[currLayer - 1]));
    }
  }

  updateNetwork(): void {
    for (let currLayer = 0; currLayer < this.layers - 1; currLayer++) {
      this.weights[currLayer] = tf.add(
        this.weights[currLayer],
        tf.mul(this.weightDerivatives[currLayer], this.learningRate)
      );

      this.biases[currLayer] = tf.add(
        this.biases[currLayer],
        tf.mul(this.biasDerivatives[currLayer], this.learningRate)
      );
    }
  }

  calculateLoss(expectedOutput: tf.Tensor): number {
    return tf.mean(
      tf.pow(tf.sub(expectedOutput, this.activations[this.layers - 1]), 2)
    ) as unknown as number;
  }
}

File: NeuralNets/Feedforward/python/numpy/NeuralNet.py
import numpy as np
import numpy.typing


class NeuralNet:
    def __init__(
        self,
        NPL: list[int],
        learning_rate: float = 0.01,
        activation_function: str = "sigmoid",
    ) -> None:
        if not NPL:
            return

        if activation_function == "sigmoid":
            self.activation_function = lambda x: 1.0 / (1.0 + np.exp(-x))
        elif activation_function == "relu":
            self.activation_function = lambda x: x if x > 0.0 else 0.0

        self.NPL: list[int] = NPL
        self.layers: int = len(self.NPL)
        self.learning_rate: float = learning_rate
        self.activations: list[numpy.typing.NDArray[np.float32]] = []
        self.weights: list[numpy.typing.NDArray[np.float32]] = []
        self.weight_derivatives: list[numpy.typing.NDArray[np.float32]] = []
        self.biases: list[numpy.typing.NDArray[np.float32]] = []
        self.bias_derivatives: list[numpy.typing.NDArray[np.float32]] = []

        for layer in range(self.layers - 1):
            self.activations.append(np.zeros(self.NPL[layer], dtype=np.float32))
            self.biases.append(np.zeros(self.NPL[layer + 1], dtype=np.float32))
            self.bias_derivatives.append(
                np.zeros(self.NPL[layer + 1], dtype=np.float32)
            )
            self.weights.append(
                np.random.uniform(
                    low=-0.3, high=0.3, size=(self.NPL[layer], self.NPL[layer + 1])
                ).astype(np.float32)
            )
            self.weight_derivatives.append(
                np.zeros((self.NPL[layer], self.NPL[layer + 1])).astype(np.float32)
            )

        self.activations.append(np.zeros(self.NPL[-1], dtype=np.float32))

    def train(
        self,
        X: numpy.typing.NDArray[np.float32],
        Y: numpy.typing.NDArray[np.float32],
        epochs: int,
    ) -> None:
        for epoch in range(epochs):
            mean_loss: float = 0.0
            for training_input, expected_output in zip(X, Y):
                self.activations[0] = training_input

                self.forward_prop()
                self.backpropagation(expected_output)

                self.update_network()
                mean_loss += self.calculate_loss(expected_output)

            mean_loss /= len(X)
            print(f"Epoch[{epoch}]: Loss = {mean_loss}")

    def forward_prop(self) -> None:
        for layer in range(self.layers - 1):
            self.activations[layer + 1] = self.activation_function(
                (self.activations[layer] @ self.weights[layer]) + self.biases[layer]
            )

    def backpropagation(self, expected: numpy.typing.NDArray[np.float32]) -> None:
        error: numpy.typing.NDArray[np.float32] = expected - self.activations[-1]

        for layer in range(self.layers - 1, 0, -1):
            delta: numpy.typing.NDArray[np.float32] = error * (
                self.activations[layer] * (1 - self.activations[layer])
            )

            self.weight_derivatives[layer - 1] = np.outer(
                self.activations[layer - 1], delta
            )
            self.bias_derivatives[layer - 1] = delta

            error = delta @ self.weights[layer - 1].T

    def update_network(self) -> None:
        for layer in range(self.layers - 1):
            self.weights[layer] += self.weight_derivatives[layer] * self.learning_rate
            self.biases[layer] += self.bias_derivatives[layer] * self.learning_rate

    def calculate_loss(self, expected: numpy.typing.NDArray[np.float32]) -> float:
        return np.mean((expected - self.activations[-1]) ** 2, dtype=float)
